In [31]:
import pandas as pd
import numpy as np

| Column | Type | Description |
|---|---|---|
| `fifa_id` | int | FIFA fantasy platform unique player identifier |
| `name` | str | Player display name from FIFA fantasy (knownName or firstName + lastName) |
| `squad_id` | int | FIFA fantasy internal squad/nation identifier |
| `team` | str | Nation name in English, mapped from squad_id |
| `position` | str | Fantasy position: DEF, MID, FWD, GK |
| `price` | float | Fantasy selection price in FIFA fantasy currency |
| `status` | str | Player availability: playing, transferred |
| `total_points` | int | Total fantasy points accumulated across all completed rounds |
| `avg_points` | float | Average fantasy points per round played |
| `matches_played` | int | Number of rounds with non-empty stat entries |
| `percent_selected` | float | Percentage of fantasy teams that have selected this player |
| `round_1_points` | int | Fantasy points earned in round 1 |
| `round_2_points` | int | Fantasy points earned in round 2 |
| `round_3_points` | int | Fantasy points earned in round 3 |
| `round_4_points` | int | Fantasy points earned in round 4 |
| `next_fixture` | int | FIFA fantasy fixture ID for the next scheduled match |

| Column | Type | Description |
|---|---|---|
| `betano_matched_name` | str | Player name as it appears in Betano scorer markets |
| `betano_match_score` | float | Fuzzy match confidence score 0 to 100; 100 indicates exact or manual override match |

| Column | Type | Description |
|---|---|---|
| `anytime_scorer_odd` | float | Betano decimal odd for player to score at any point; defaults to 150.0 if team has a match but player has no Betano entry |
| `anytime_scorer_prob` | float | Raw implied probability from anytime scorer odd (1 / odd) |
| `first_scorer_odd` | float | Betano decimal odd for player to score the first goal of the match |
| `first_scorer_prob` | float | Raw implied probability from first scorer odd (1 / odd) |
| `last_scorer_odd` | float | Betano decimal odd for player to score the last goal of the match |

| Column | Type | Description |
|---|---|---|
| `match_home` | str | Home team name for the player's next match |
| `match_away` | str | Away team name for the player's next match |
| `is_home` | bool | True if the player's team is the home side |
| `opponent` | str | Opposing team name |
| `match_home_win_odd` | float | Betano decimal odd for home team victory |
| `match_draw_odd` | float | Betano decimal odd for a draw |
| `match_away_win_odd` | float | Betano decimal odd for away team victory |
| `match_btts_prob` | float | De-vigged probability that both teams score |
| `match_over_25_odd` | float | Betano decimal odd for over 2.5 total goals |
| `match_over_25_prob` | float | De-vigged probability of over 2.5 total goals |
| `match_over_35_odd` | float | Betano decimal odd for over 3.5 total goals |
| `team_over_05_odd` | float | Betano decimal odd for the player's team to score at least 1 goal |
| `team_over_15_odd` | float | Betano decimal odd for the player's team to score at least 2 goals |
| `team_score_prob` | float | De-vigged probability the player's team scores at least once |
| `team_score_2_prob` | float | Implied probability the player's team scores 2 or more goals (1 / team_over_15_odd) |
| `team_cs_prob` | float | De-vigged probability the player's team keeps a clean sheet |
| `team_qualify_odd` | float | Betano decimal odd for the player's team to advance to the next round |
| `team_win2_odd` | float | Betano decimal odd for the player's team to win by 2 or more goals |
| `opp_over_05_odd` | float | Betano decimal odd for the opponent to score at least 1 goal |
| `opp_score_prob` | float | De-vigged probability the opponent scores at least once |
| `opp_cs_prob` | float | De-vigged probability the opponent keeps a clean sheet |

| Column | Type | Description |
|---|---|---|
| `stat_GS` | int | Goals scored during the tournament (FIFA fantasy stat) |
| `stat_AS` | int | Assists during the tournament (FIFA fantasy stat) |
| `stat_CS` | int | Clean sheets during the tournament (FIFA fantasy stat) |
| `stat_GC` | int | Goals conceded during the tournament (FIFA fantasy stat) |
| `stat_MP` | int | Minutes played during the tournament (FIFA fantasy stat) |
| `stat_YC` | int | Yellow cards during the tournament (FIFA fantasy stat) |
| `stat_RC` | int | Red cards during the tournament (FIFA fantasy stat) |
| `stat_ST` | int | Shots on target during the tournament (FIFA fantasy stat) |
| `stat_SB` | int | Shots blocked during the tournament (FIFA fantasy stat) |
| `stat_CC` | int | Chances created during the tournament (FIFA fantasy stat) |
| `stat_PS` | int | Penalties saved during the tournament (FIFA fantasy stat, GK) |
| `stat_T` | int | Tackles during the tournament (FIFA fantasy stat) |
| `stat_S` | int | Saves during the tournament (FIFA fantasy stat, GK) |
| `stat_SXI` | int | Number of times named in the starting XI (FIFA fantasy stat) |
| `stat_OG` | int | Own goals during the tournament (FIFA fantasy stat) |
| `stat_PC` | int | Penalties committed during the tournament (FIFA fantasy stat) |
| `stat_PW` | int | Penalties won during the tournament (FIFA fantasy stat) |
| `stat_FK` | int | Free kicks earned during the tournament (FIFA fantasy stat) |


| Column                  | Type  | Description                                                         |
| ----------------------- | ----- | ------------------------------------------------------------------- |
| `club_MP`               | int   | Matches played in top 5 European leagues, 2025/26 season            |
| `club_Starts`           | int   | Matches started in top 5 European leagues                           |
| `club_Min`              | int   | Total minutes played in top 5 European leagues                      |
| `club_90s`              | float | Minutes played expressed as full 90-minute equivalents              |
| `club_Gls`              | int   | Goals scored in top 5 European leagues                              |
| `club_Ast`              | int   | Assists in top 5 European leagues                                   |
| `club_G+A`              | int   | Goals plus assists                                                  |
| `club_G-PK`             | int   | Non-penalty goals                                                   |
| `club_TklW`             | int   | Tackles won                                                         |
| `club_Int`              | int   | Interceptions                                                       |
| `club_GA`               | int   | Goals conceded (GK only)                                            |
| `club_Saves`            | int   | Saves (GK only)                                                     |
| `club_CS`               | int   | Clean sheets (GK only)                                              |
| `club_PKA`              | int   | Penalties faced (GK only)                                           |
| `club_PKsv`             | int   | Penalties saved (GK only)                                           |
| `club_CrdY`             | int   | Yellow cards                                                        |
| `club_CrdR`             | int   | Red cards                                                           |
| `club_PK`               | int   | Penalties scored                                                    |
| `club_PKatt`            | int   | Penalty attempts                                                    |
| `club_SoTA`             | int   | Shots on target faced (GK only)                                     |
| `club_PKm`              | int   | Penalties missed                                                    |
| `club_Sh`               | int   | Shots                                                               |
| `club_SoT`              | int   | Shots on target                                                     |
| `club_Fls`              | int   | Fouls committed                                                     |
| `club_Fld`              | int   | Fouls drawn                                                         |
| `club_Off`              | int   | Offsides                                                            |
| `club_Crs`              | int   | Crosses                                                             |
| `club_OG`               | int   | Own goals                                                           |
| `club_2CrdY`            | int   | Second yellow cards                                                 |
| `club_Save%`            | float | Save percentage (GK); unreliable if multi-club                      |
| `club_CS%`              | float | Clean sheet percentage; unreliable if multi-club                    |
| `club_GA90`             | float | Goals conceded per 90 (GK)                                          |
| `club_SoT%`             | float | Shots on target %                                                   |
| `club_Sh/90`            | float | Shots per 90                                                        |
| `club_SoT/90`           | float | Shots on target per 90                                              |
| `club_G/Sh`             | float | Goals per shot                                                      |
| `club_G/SoT`            | float | Goals per shot on target                                            |
| `club_Pos`              | str   | Position code (FW, MF, DF, GK)                                      |
| `club_Squad`            | str   | Clubs played for (slash-separated if multiple)                      |
| `club_Comp`             | str   | Competitions played in                                              |
| `club_Age`              | int   | Player age                                                          |
| `club_fbref_multi_club` | bool  | True if multiple clubs in season; rate stats unreliable if True     |
| `club_matched_player`   | str   | FBref player name matched via fuzzy match                           |
| `club_match_dist`       | int   | Edit distance between names (0 = exact match, i.e. 100% confidence) |



In [32]:
df= pd.read_csv("fantasy_enriched.csv")

In [33]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 625 entries, 0 to 624
Columns: 106 entries, fifa_id to clubs_match_dist
dtypes: bool(1), float64(88), int64(4), object(1), str(12)
memory usage: 565.8+ KB


In [34]:
print(f"Dataset shape: {df.shape[0]:,} rows x {df.shape[1]} columns")
print("\n".join(df.columns))

Dataset shape: 625 rows x 106 columns
fifa_id
name
squad_id
team
position
price
status
total_points
avg_points
matches_played
percent_selected
round_1_points
round_2_points
round_3_points
round_4_points
next_fixture
betano_matched_name
betano_match_score
anytime_scorer_odd
anytime_scorer_prob
first_scorer_odd
first_scorer_prob
last_scorer_odd
match_home
match_away
is_home
opponent
match_home_win_odd
match_draw_odd
match_away_win_odd
match_btts_prob
match_over_25_odd
match_over_25_prob
match_over_35_odd
team_over_05_odd
team_over_15_odd
team_score_prob
team_score_2_prob
team_cs_prob
team_qualify_odd
team_win2_odd
opp_over_05_odd
opp_score_prob
opp_cs_prob
stat_GS
stat_AS
stat_CS
stat_GC
stat_MP
stat_YC
stat_RC
stat_ST
stat_SB
stat_CC
stat_PS
stat_T
stat_S
stat_SXI
stat_OG
stat_PC
stat_PW
stat_FK
club_MP
club_Starts
club_Min
club_90s
club_Gls
club_Ast
club_G+A
club_G-PK
club_TklW
club_Int
club_GA
club_Saves
club_CS
club_PKA
club_PKsv
club_CrdY
club_CrdR
club_PK
club_PKatt
club_SoTA
club_

In [35]:
print(df.shape)

(625, 106)


In [36]:
df.isnull().sum().sum()

np.int64(16985)

In [37]:
categorical_cols = df.select_dtypes(include="object").columns.tolist()
numerical_cols = df.select_dtypes(include=np.number).columns.tolist()

/tmp/ipykernel_277464/2784395380.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include="object").columns.tolist()


In [38]:
# Table for missing values distribution and percentage.
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
missing_df = pd.DataFrame({"n_missing": missing, "%_missing": missing_pct})
missing_df = missing_df[missing_df["n_missing"] > 0].sort_values("n_missing", ascending=False)

pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

missing_df

,n_missing,%_missing
club_GA90,603,96.48
club_CS%,603,96.48
club_Save%,603,96.48
round_4_points,564,90.24
club_G/SoT,358,57.28
club_SoT%,340,54.40
club_G/Sh,340,54.40
club_Ast,312,49.92
club_Gls,312,49.92
club_G-PK,312,49.92


missing values in club at 98% is anyone that is either not a goalkeeper or not in the stats sheet
club_G/SoT	358	57.28
club_SoT%	340	54.40
club_G/Sh	340	54.40 these slightly higher than the others i assume are defenders and others who have not had goals or shots or cannot have a value assingned to this metric is like 20 rows more than the common mnar
all the club 49.92 are players playing in the world cup outside top 5 leagues number will change based on left over teams ofc
round missing values are players who have not played / entered minutes. so 0, but different than a 0 from playing but not making points.
betano scorers names are goalkeepers, very unkown players players who will not even play basically goal keepers and the back of the row in teams national teams
all betano names are assinged so these are unlisted players. 
the team prob is weird betano fetcher missed some odds 